src/ILCompiler.Compiler/src/Compiler/PreInitFieldInfo.cs

```diff
@@ -199,26 +202,37 @@ private static FieldDesc GetPreInitDataField(FieldDesc thisField)
         /// </summary>
         private static PreInitFieldInfo ConstructPreInitFieldInfo(FieldDesc field, FieldDesc dataField)
         {
-            var arrType = field.FieldType as ArrayType;
-            if (arrType == null || !arrType.IsSzArray)
-            {
-                // We only support single dimensional arrays
-                throw new NotSupportedException();
-            }
-
             if (!dataField.HasRva)
                 throw new BadImageFormatException();
-            
+
             var ecmaDataField = dataField as EcmaField;
             if (ecmaDataField == null)
                 throw new NotSupportedException();
-            
+
             var rvaData = ecmaDataField.GetFieldRvaData();
 
-            int elementSize = arrType.ElementType.GetElementSize().AsInt;
-            if (rvaData.Length % elementSize != 0)
-                throw new BadImageFormatException();
-            int elementCount = rvaData.Length / elementSize;
+            var fieldType = field.FieldType;
+
+            int elementCount;
+            if (fieldType.IsValueType || fieldType.IsPointer)
```

Comment:
I would structure this code as:

```csharp
if (fieldType.IsValueType || fieldType.IsPointer)
...
else if (fieldType.IsSzArray)
...
else
    throw
```

The current structure shows too much of the code's history and is pretty unreadable as a result.